File: easy/1351. Count Negative Numbers in a Sorted Matrix/main.go
package main

import "math"

func binSearchNegative(arr []int) int {
	left, right, prev, idx := 0, len(arr)-1, math.MaxInt, -1

	for left <= right {
		mid := left + (right-left)/2
		if arr[mid] >= 0 {
			left = mid + 1
			if arr[mid] <= prev {
				prev = arr[mid]
				idx = mid
			}
		} else {
			right = mid - 1
		}
	}

	if idx == -1 {
		return len(arr)
	}

	return len(arr) - (idx + 1)
}

func countNegatives(grid [][]int) int {
	nums := 0
	for _, row := range grid {
		nums += binSearchNegative(row)
	}

	return nums
}

File: easy/1351. Count Negative Numbers in a Sorted Matrix/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/count-negative-numbers-in-a-sorted-matrix/

const searchNegativeInRow = (row: number[]): number => {
  let start = 0,
    end = row.length - 1,
    mid: number,
    prev: number = Number.MAX_VALUE,
    ans: number = -1
  while (start <= end) {
    mid = start + Math.floor((end - start) / 2)
    if (row[mid] >= 0) {
      start = mid + 1
      if (row[mid] <= prev) {
        prev = row[mid]
        ans = mid
      }
    } else if (row[mid] < 0) {
      end = mid - 1
    }
  }

  return ans == -1 ? row.length : row.length - (ans + 1)
}

const countNegatives = (grid: number[][]): number => {
  let count = 0
  for (let row of grid) {
    count += searchNegativeInRow(row)
  }

  return count
}
